### PR TITLE
Add store/forward cache capabilities

### LINKED ISSUE #3: Provide store/forward cache capabilities
**Is your feature request related to a problem? Please describe.**
The cache requires constant connectivity.  Please add a cache with store/forward features so the proxy can run both offline and online.

**Describe the solution you'd like**
The cache should use a local single-file database so that it can easily be transferred from a behind-the-firewall instance of the proxy (running in stored mode) to an outside-the-firewall instance (running in forward mode) and then back again. 
(Even better than full file transfer would be to support import and export of requests and responses from one database file to another.)

### PR DESCRIPTION
**Pull request by adobeDan**
## Summary
* Add an sqlite cache for both activation and deactivation requests
* Respond from the cache if possible when COPS is unavailable
* Provide for store (cache-only) and forward (COPS-only) proxy operating modes

## Testing Steps
* Test in passthrough mode (for SSL, you will need to update your cert to pkcs12 format with a password)
* Test in cache mode
* Test in store mode
* Test in forward mode

Fixes #3

### CODE REVIEW

**Comment by adobeDan** [comment]
@adorton-adobe this is FYI about a lot of work on caching I did over the break.  I still need to get sqlx working the latest version of tokio before I can fully test the cache.  I should be done with this in about a week.  Then I would be happy to turn it over to one of your folks.

**Comment by adobeDan** [comment]
Hey, @adorton-adobe, it's definitely time to review this PR, because I believe it works great!

**Comment by adobeDan** [comment]
I'm attaching a [working config file and certificate](https://github.com/adobe/frl-online-proxy/files/5813030/config.zip) to make it easier to test.  Just add these lines to your local host file:
```
## section: SSL
# certificate to use with localhost
127.0.0.1	frl-proxy.brotsky.net
::1			frl-proxy.brotsky.net
## end of section: SSL
``` 
and it should work fine.  You can control the mode with the --mode command line argument, as in
```
frl-online-proxy -c config.toml --mode store
```

**Comment by adorton-adobe** [comment]
Thanks. I started looking at it yesterday.

File: src/cops.rs
Comment (adorton-adobe) [inline src/cops.rs:175]: When I test license deactivation, I get a URL parse error - `RelativeUrlWithoutBase`. The URL that we get from the request parts appears to just include the path and query string. Example:

```
/asnp/frl_connected/v1?npdId=NTU4YThhYz...mE4YmQ1YzA1&deviceId=d5a9c61202a6a0a9...1c7681714f1b&osUserId=3b831e95ca3c1802aa687c...3a70cbc&enableVdiMarkerExists=0&isVirtualEnvironment=0&isOsUserAccountInDomain=1
```

Without the host component, the `url` crate is unable to parse the query string. One option is to add the hostname from `Settings`, but we're not getting the conf object here. It seems sufficient just to put in some placeholder. Here's a quick fix I came up with and tested successfully:

```rust
        let request_url = format!("http://placeholder{}", &parts.uri.to_string());
        let pairs: HashMap<String, String> = Url::parse(&request_url)
            .expect("Bad deactivation query string")
            .query_pairs()
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .collect();
```

**Review by adorton-adobe — CHANGES_REQUESTED** [review]
Looks great. I just have a few comments including one bug I encountered. Once we get those cleared up I think we're ready to merge. I will continue to test it in the meantime.

**Comment by adorton-adobe** [comment]
One other thing I forgot to ask - have you made any updates to the User Guide yet? If not, I will plan to update it to cover this new functionality.

File: Cargo.toml
Comment (jplatte) [reply Cargo.toml:32]: FWIW, I usually delete branches after they've been merged. I would not consider anything in the fork stable.

Comment (adorton-adobe) [reply Cargo.toml:32]: Gotcha, thank you.

@adobeDan - I've forked the repo I suggested above - let's use `tokio-1.0` on that until we get the update in a release.

```
https://github.com/adorton-adobe/sqlx
```

File: src/cops.rs
Comment (adobeDan) [reply src/cops.rs:175]: I'm glad you tested deactivation - I hadn't yet so it doesn't surprise me there are lurking bugs.  I'll take a look and probably use your fix.

File: Cargo.toml
Comment (adobeDan) [reply Cargo.toml:32]: I completely agree that we need our own pet repo holding the hacked sqlx, and I will use yours for now.  It looks like the sqlx folks are actually going to release a beta that has this code in it; that would be more reliable and allow us to use a standard release tag instead of our own repo.

File: src/cli.rs
Comment (adobeDan) [reply src/cli.rs:22]: Yep, great point.  I got lazy because it was a change in direction to use a string; originally I had an enum but those don't work well with structopt and settings.

**Comment by adobeDan** [comment]
> One other thing I forgot to ask - have you made any updates to the User Guide yet? If not, I will plan to update it to cover this new functionality.

No, I hate doing docs, I'm slow at it and overthink them terribly.  If you would be willing that would be great.

**Comment by adobeDan** [comment]
@adorton-adobe I've integrated your requested fixes, and I've tested deactivation in cache, store, and forward modes.  I think we're ready to merge unless your testing reveals other problems.

**Review by adorton-adobe — APPROVED** [review]
I think we're good for now. I will be doing additional testing as I update the docs and will open issues for anything I encounter.

**Pull request merged**
